### PR TITLE
DEV: Allow SiteSetting::Update service to bulk update

### PR DESCRIPTION
**Pull request by lis2**
Previously, the SiteSetting::Update service allowed to update of a single site setting. In the About controller, we were using the loop through all settings - https://github.com/discourse/discourse/blob/main/app/controllers/admin/config/about_controller.rb#L39

It is suboptimal because if the 3 first settings are saved and the fourth is invalid, we will end with partially updated data.

Changing SiteSetting::Update to accept hash means that we will check upfront if none of the settings are hidden or invalid and update all or none.

Custom policies are used to report which settings are failing.

### CODE REVIEW

File: app/services/site_setting/update.rb [hunk 1]
```diff
@@ -47,21 +54,9 @@ def current_user_is_admin(guardian:)
...
-
-  def setting_is_configurable(params:)
-    return true if !SiteSetting.plugins[params.setting_name]
-
-    Discourse.plugins_by_name[SiteSetting.plugins[params.setting_name]].configurable?
-  end
-
   def save(params:, guardian:)
-    SiteSetting.set_and_log(params.setting_name, params.new_value, guardian.user)
+    params.settings.each do |setting_name, value|
```
Comment (martin-brennan) [inline app/services/site_setting/update.rb:58]: I think a small improvement would be to run this in a transaction block, so either all settings save or none do

**Review by martin-brennan — APPROVED** [review]
Very nice, love the custom policy class usage :ok_hand:

File: app/services/site_setting/update.rb [hunk 2]
```diff
@@ -3,42 +3,49 @@
...
-          new_value.blank? ? "" : Upload.get_from_urls(new_value.split("|")).to_a
-        when :upload
-          Upload.get_from_url(new_value) || ""
-        else
-          new_value
-        end
+      self.settings =
+        self
+          .settings
+          .to_a
```
Comment (Flink) [inline app/services/site_setting/update.rb:23]: `settings` should be a hash here because of the `before_validation` block, so a call to `#to_a` shouldn’t be necessary.

**Review by Flink — APPROVED** [review]
Some small comments/suggestions, but great improvement!

File: spec/services/site_setting/update_spec.rb [hunk 3]
```diff
@@ -84,5 +85,11 @@
         end
       end
     end
+
+    context "when one setting is having invalid value" do
```
Comment (OsamaSayegh) [inline spec/services/site_setting/update_spec.rb:89]: Can we extend this further and add a test case for when multiple settings are updated and one of them fails and ensure that none of the settings are actually updated.

**Review by OsamaSayegh — APPROVED** [review]
